server/src/auth/index.ts

```diff
@@ -202,3 +229,34 @@ export function ensureFileGroupAccess() {
     }
   };
 }
+
+/**
+ * Middleware function to protect published submissions that require access. Sets
+ * included materials to the res.locals property so they can be accessed also later
+ * @returns Request middleware
+ */
+export function ensurePublicationAccess() {
+  return async (req: Request, res: Response, next: NextFunction) => {
+    const user = basicAuth(req);
+    const surveyId = Number(req.params.id);
+    let authorized = false;
+
+    if (user) {
+      const accesses = await getPublicationAccesses(
+        surveyId,
+        user.name,
+        user.pass,
+      );
+      res.locals.alphanumericIncluded = accesses.alphanumericIncluded;
+      res.locals.geospatialIncluded = accesses.geospatialIncluded;
+      res.locals.personalIncluded = accesses.personalIncluded;
+      authorized = accesses.authorized;
+    }
+    if (!authorized) {
+      res.set('WWW-Authenticate', 'Basic realm="example"');
```

Comment:
The realm should be more descriptive - on the credentials prompt the client may tell the realm, so maybe this should just be in Finnish? It should probably also include the survey ID, so people authenticating to multiple survey answer publications wouldn't have their credentials mixed up - maybe something like this?

```suggestion
      res.set('WWW-Authenticate', `Basic realm="Kyselyn ${surveyId} vastaukset"`);
```